### PR TITLE
Better SDK34 support

### PR DESCRIPTION
**Pull request by TheRealAgentK**
### Description :memo:
This PR is to add **SDK 34 compatibility** to the raygun4android provider. We had a user coming through mentioning that they can't use raygun4android at all on their project, due to the forward compatibility not working well. So we've decided to jut update the SDK to version 34, meeting the project requirement. 

**Type of change**
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Updates**
- Makes both `:provider` and `:app` modules work correctly with SDK34 and with more modern dependencies
- Fixed Timber documentation (#69)
- Fixed Jetifier dependency along the way (#83)
- Fix duplicate naming issue (#87)

### Test plan :test_tube:
- Import project
- build `:provider` from clean, build `:app`
- Launch app in emulator

### Author to check :eyeglasses:
- [x] Project and all contained modules build
- [x] Tested on appropriate devices/emulators and SDK versions
- [x] Reviewed by another developer
- [x] Appropriate documentation written (code comments, internal docs)

### Reviewer to check :heavy_check_mark:
- [x] Change has been tested on a device/emulator
- [x] Code is written to standards
~- [ ] Appropriate tests have been written (code comments, internal docs)~ - N/A right now

### Caveats:
- Build needs to be updated to Gradle 8 to be "officially supported" with SDK34. Separate PR to follow.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,7 +47,7 @@ Then add the following to your **module's** build.gradle:
 ```gradle
 dependencies {
     ...
-    implementation 'com.raygun:raygun4android:4.+'
+    implementation 'com.raygun:raygun4android:4.0.1'
```
Comment (sumitramanga) [inline README.md:50]: What's the reason behind this change?

File: README.md [hunk 2]
```diff
@@ -88,11 +90,11 @@ The above exception handler automatically catches and sends all uncaught excepti
 
 For an actual usage example, check out the sample application in the **app** module of this project
 
-## Raygun and ProGuard
+## Raygun and ProGuard/R8
 
 ### General
 
-ProGuard is a free Java tool for obfuscation, class file shrinking, optimizing and pre-verifying. When enabling ProGuard in a native Android application that also uses Raygun, the obfuscation feature requires a bit of attention. By default, your obfuscated class and method names will show up in the stacktraces of exception/error reports submitted to Raygun. This makes the stacktraces difficult to read when looking into the cause of the issues.
+ProGuard and R8 are tools for obfuscation, class file shrinking, optimizing and pre-verifying. When enabling ProGuard in a native Android application that also uses Raygun, the obfuscation feature requires a bit of attention. By default, your obfuscated class and method names will show up in the stacktraces of exception/error reports submitted to Raygun. This makes the stacktraces difficult to read when looking into the cause of the issues.
```
Comment (sumitramanga) [inline README.md:97]: Is it no longer free or is R8 not free bu ProGuard is?

File: README.md [hunk 3]
```diff
@@ -174,7 +176,7 @@ def createRaygunNotifyDeploymentTask(token,key,groupName,version,userName,userEm
 
 This function gets called from within the ```android {...}``` block of the Gradle file at each build in Android Studio and creates the appropriate parameterised task to notify the Raygun backend of your app's deployment.
 
-## Advanced features
```
Comment (sumitramanga) [inline README.md:177]: We mention Advanced features in most of our provider's documentation. For consistency and familiarity across the providers, we should keep this in place

File: README.md [hunk 4]
```diff
@@ -214,7 +216,7 @@ Please note that setting a custom endpoint will stop Crash Report or Real User M
 
 ### Storing crash reports on the device
 
-If the device can't connect, Raygun4Android will save the crash report to disk. At the next start of the application (and therefore the provider) it will check if the internet is now available, and if it is, send the cached messages. A maximum of 64 messages will be cached and you can change the amount by calling:
+If the device can't connect, Raygun4Android will save the crash report to storage. At the next start of the application (and therefore the provider) it will check if the internet is now available, and if it is, send the cached messages. A maximum of 64 messages will be cached and you can change the amount by calling:
```
Comment (sumitramanga) [inline README.md:219]:
```suggestion
If the device can't connect, Raygun4Android will save the crash report to the device storage. At the next start of the application, (along with the provider) it will check if the internet is now available. If it is, send the cached messages. A maximum of 64 messages will be cached and you can change the amount by calling:
```

File: README.md [hunk 5]
```diff
@@ -379,26 +380,26 @@ Provides an instance of a class which has an onBeforeSend method that can be use
 
 * Is there an example app?
 
-  Yup - clone this repository then run the **app** module of the project.
+  Yup - clone this repository, then run the **app** module of the project.
 
-* Not seeing errors in the dashboard?
+* I'm not seeing errors in the dashboard.
```
Comment (sumitramanga) [inline README.md:385]:
```suggestion
* I'm not seeing errors in Raygun Crash Reporting.
```

Comment (sumitramanga) [reply README.md:385]: Because referring to the Crash Reporting product as a dashboard is not something we do anymore. We have an actual product called Dashboards so this will be confusing to keep!

File: README.md [hunk 6]
```diff
@@ -379,26 +380,26 @@ Provides an instance of a class which has an onBeforeSend method that can be use
...
 
   Raygun4Android outputs Logcat messages - look for the the logcat tag **Raygun4Android**. HTTP Status 403 will indicate an invalid API key, 400 a bad message, and 202 will indicate received successfully.
 
 * My build fails with `Default interface methods are only supported starting with Android N (--min-api 24)`. Why is that?
 
-  Raygun4Android uses Timber for internal logging. This requires some language features that are only available with Java 8. Make sure that your project using the library has set compilation compatibility to Java 8.
+  Raygun4Android uses Timber for internal logging. This requires some language features that are only available with Java 8. Make sure that your project using the library has set compilation compatibility to Java 8 as the minimum
```
Comment (sumitramanga) [inline README.md:391]:
```suggestion
  Raygun4Android uses Timber for internal logging. This requires some language features that are only available with Java 8. Make sure that your project, using the library, has set compilation compatibility to Java 8 as the minimum
```

File: README.md [hunk 7]
```diff
@@ -415,3 +416,17 @@ android {
 * The library logs an error message about a not found class: Rejecting re-init on previously-failed class java.lang.Class<android.support.v4.app.JobIntentService$JobServiceEngineImpl>
 
   The message above stems from certain versions of the Android support libraries. JobServiceEngineImpl is part of Android Oreo (8, SDK 26) and newer only. The support library catering for supporting services on earlier versions of Android runs internal checks for which implementation is available to it. As part of the checks, it outputs the message as an informational feedback only.
+
+* Timber Lint warnings get exposed to my app even though I don't (want to) use Timber.
+
+  The solution for the time being is to disable linting for those specific lint warning in your app:
```
Comment (sumitramanga) [inline README.md:422]: Can we say for production? because it will helpful for the developer locally but not in production towards the app users

File: app/build.gradle [hunk 8]
```diff
@@ -75,14 +75,17 @@ android {
             storePassword '123456'
         }
     }
-    compileSdk 28
+    compileSdk 34
```
Comment (sumitramanga) [inline app/build.gradle:78]: Woohoo 💪🏽

File: README.md (hunk 1 above)
Comment (TheRealAgentK) [reply README.md:50]: I want to tell people that at this stage - until anything else is published - 4.0.1 is the stable released. `4.+` might just unintendend suck in a 4.1.0 alpha/snapshot when I publish it.

File: README.md (hunk 2 above)
Comment (TheRealAgentK) [reply README.md:97]: The `free` comment doesn't really make sense in this context and is probably still in the doc from before I even took it over :). It comes with Android Studio as-is anyway and is now - depending on what AS folks use - either ProGuard or R8.

File: app/build.gradle (hunk 8 above)
Comment (TheRealAgentK) [reply app/build.gradle:78]: 😸

File: README.md (hunk 7 above)
Comment (TheRealAgentK) [reply README.md:422]: Just to clarify - the users of an app will never see this.

What happens is:

- `Raygun4Android:provider` uses Timber (a library for better logging on the device's logcat)
- Timber comes with its own set of Android (Studio) Lint rules and they're opinionated.
- Due to how Android's toolchain works, the linting rules are passed "downstream" to any modulde/app that uses a library (our `:provider`) that uses Timber. In this case to `Raygun4Android:app`.
- So, anything that *uses* our `:provider` gets the Timber linting rules passed through to it, regardless if that app is using Timber itself or not (many apps do anyway).
- Effectively that means that an app developer not using Timber in their app sees linting rules and might be confused. From here the options for them are to switch off linting in their IDE completely or ignore the lint rules in their code etc - but the cleanest solution is currently to add this disable statement into the build.

So, this is not inherently a `debug` or `release` issue. In their app, they should do this anyway if they don't use Timber themselves.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheRealAgentK) [reply README.md]: If the storage limit (with the default of 64) is reached, it will just not write more unless there is space. If that's a good way of doing it, I'm not certain - but it was at the time aligning how Mitchell had written the iOS provider, so we tried to keep the behaviour in line.

I think the default of 64 is not really that reasonable anymore these days either - we should consider to change that maybe. There's also https://github.com/MindscapeHQ/raygun4android/issues/49 - that goes into that direction of "we should really consider changing that" as well.

File: app/build.gradle [hunk 9]
```diff
@@ -75,14 +75,17 @@ android {
...
         minSdkVersion 16
-        targetSdkVersion 28
+        targetSdkVersion 34
         versionCode = Long.valueOf(VERSION_CODE)
         versionName = "${VERSION_NAME}-sample"
-        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
+    }
+    lintOptions {
+        disable 'LogNotTimber','StringFormatInTimber', 'ThrowableNotAtBeginning', 'BinaryOperationInTimber', 'TimberArgCount', 'TimberArgTypes', 'TimberTagLength'
```
Comment (sumitramanga) [inline app/build.gradle:88]: Just to confirm - We are excluding these lint options for now?

Comment (TheRealAgentK) [reply app/build.gradle:88]: Yes, in the `:app` module. Because the sample app doesn't use Timber itself and we don't want the `:provider`'s Timber inclusion trigger lint warnings in the app that are not relevant. And it's a great sample of how to work around the toolchain's flow.

File: app/build.gradle [hunk 10]
```diff
@@ -108,21 +111,22 @@ android {
     // Setup RG tasks
     createRaygunNotifyDeploymentTask(RAYGUN_API_TOKEN, RAYGUN_API_KEY, RAYGUN_GROUP, defaultConfig.versionName, localUserName, localUserEmail)
     createRaygunProguardTask(RAYGUN_API_TOKEN, RAYGUN_APP_PATH, RAYGUN_GROUP, defaultConfig.versionName)
-    namespace 'com.raygun.raygun4androidsample.sample'
+    namespace 'com.raygun.raygun4android.sample'
```
Comment (sumitramanga) [inline app/build.gradle:114]: Thank you for fixing this up!

File: app/src/debug/AndroidManifest.xml
Comment (sumitramanga) [inline app/src/debug/AndroidManifest.xml:1]: Is this file meant to be committed?

File: app/src/main/java/com/raygun/raygun4android/sample/MainActivity.kt [hunk 11]
```diff
@@ -91,8 +91,8 @@ class MainActivity : AppCompatActivity() {
         }
 
 
-        textViewAppVersion.text = "App ${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE} ${BuildConfig.BUILD_TYPE})"
-        textViewProviderVersion.text = "Provider ${com.raygun.raygun4android.BuildConfig.VERSION_NAME} (${com.raygun.raygun4android.BuildConfig.VERSION_CODE} ${BuildConfig.BUILD_TYPE})"
+        textViewAppVersion.text = getString(R.string.app_version_text, BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE, BuildConfig.BUILD_TYPE)
+        textViewProviderVersion.text = getString(R.string.provider_version_text, BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE, BuildConfig.BUILD_TYPE)
```
Comment (sumitramanga) [inline app/src/main/java/com/raygun/raygun4android/sample/MainActivity.kt:95]: This tidy up looks much better!

File: app/src/debug/AndroidManifest.xml
Comment (TheRealAgentK) [reply app/src/debug/AndroidManifest.xml:1]: Yup

Comment (TheRealAgentK) [reply app/src/debug/AndroidManifest.xml:1]: To explain a bit more:

An Android build has a `debug` and `release` build process. The first is usually used when people work on an app and it would - depending on their setup - then also bring in debug builds of libraries. 

So this makes an addition for the debug build of the sample app's Manifest. The line in there adds a permission a - debug-build-only - tooling library (leakcanary) requires on new Android releases. This files gets merged together with the main `AndroidManifest` in the build so that debug build artefacts have that permission while release artefacts don't.

File: build.gradle [hunk 12]
```diff
@@ -1,7 +1,7 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
-    ext.kotlin_version = '1.9.0'
+    ext.kotlin_version = '1.8.22'
```
Comment (sumitramanga) [inline build.gradle:4]: Is it possible that the AGP 7.4.1 needs to be upgraded in order for this to not be downgraded?

File: provider/src/main/res/values/strings.xml
Comment (sumitramanga) [inline provider/src/main/res/values/strings.xml:1]: Is there are no resources, can this file be deleted?

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheRealAgentK) [reply README.md]: Just pushed an updated with a note on the current behaviour.

File: README.md (hunk 3 above)
Comment (TheRealAgentK) [reply README.md:177]: Fixed

File: build.gradle (hunk 12 above)
Comment (TheRealAgentK) [reply build.gradle:4]: Yeah, I had it down to 1.7.x as part of the process for a while and then ran into other issues. It'll change again when I do the Gradle 8 change.

File: provider/src/main/res/values/strings.xml
Comment (TheRealAgentK) [reply provider/src/main/res/values/strings.xml:1]: That's an interesting question because the file gets created by default when you setup a project and has `app_name` in there, in case you want to use it - I guess. We didn't use the resource in the provider, so I emptied it. I'd be cautious with removing it - but I'll give that a try in the next piece of work. Right now it doesn't really bother me being there.

**Review by sumitramanga — APPROVED** [review]
LGTM! 💯 Would you like me to test on my end or are you sweet? @TheRealAgentK

**Comment by sumitramanga** [comment]
I get what looks like warnings when pulling down the branch and cleaning the project - 
image

Does this happen on your end?

**Comment by sumitramanga** [comment]
One of the warnings look related to Gradle upgrade which yo mentioned would be done in a separate PR. That can be ignored for now

**Comment by TheRealAgentK** [comment]
> I get what looks like warnings when pulling down the branch and cleaning the project - image
> 
> Does this happen on your end?

Yeah, that's because of Gradle deprecations for 8.

**Comment by sumitramanga** [comment]
The build itself succeeded on my local machine 👍🏽

File: provider/src/main/res/values/strings.xml
Comment (TheRealAgentK) [reply provider/src/main/res/values/strings.xml:1]: #94
